### ISSUE TITLE
Should configure PodDisruptionBudgets to defend replicas from node drains

### ISSUE THREAD

**Issue by anguslees** [rook/rook#2128]
**Is this a bug report or feature request?**
* Bug Report

I consider lack of this configuration to be a bug, since it is a necessary part of operating a stateful service in a k8s environment - but I understand if someone wants to re-label this as a new feature instead.

**Deviation from expected behavior:**

I have a basic n+1 host redundancy rook configuration (ie: replicated pool with failureDomain=host and size=2; mon count=3).  `kubectl drain` allows me to drain more than one node containing an osd or mon pod, making data unavailable during this period.

**Expected behavior:**

`kubectl drain` on the second node should block until the first node is undrained and osd/mon pod is healthy again.  More generally, I shouldn't be allowed to drain as many nodes as there are pool replicas, or a majority of mon pods, or all the mds replicas.

iow: I think rook should have configured PodDisruptionBudgets that look something like this:
```yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
spec:
  maxUnavailable: 1    # pool.spec.replicated.size - 1
  selector:
    matchLabels:
      app: rook-ceph-osd
      rook_cluster: rook-ceph   # as appropriate
---
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
spec:
  minAvailable: 2   # floor(cluster.spec.mon.count / 2 + 1)
  selector:
    matchLabels:
      app: rook-ceph-mon
      mon_cluster: rook-ceph   # as appropriate
---
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
spec:
  minAvailable: 1
  selector:
    matchLabels:
      app: rook-ceph-mds
      rook_cluster: rook-ceph   # as appropriate
```

**How to reproduce it (minimal and precise):**

```yaml
apiVersion: ceph.rook.io/v1beta1
kind: Cluster
spec:
  mon:
    allowMultiplePerNode: false
    count: 3
# etc
---
apiVersion: ceph.rook.io/v1beta1
kind: Pool
spec:
  failureDomain: host
  replicated:
    size: 2
```

```sh
kubectl drain --ignore-daemonsets $node1
kubectl drain --ignore-daemonsets $node2
```

Then observe data replica (un)availability.

**Environment**:
* Rook version (use `rook version` inside of a Rook Pod):
` rook: v0.8.1`

* Kubernetes version (use `kubectl version`):
`Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.15", GitCommit:"c2bd642c70b3629223ea3b7db566a267a1e2d0df", GitTreeState:"clean", BuildDate:"2018-07-11T17:52:15Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/arm"}`

* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
Baremetal/DIY

**Comment by anguslees** [comment]
@galexrt I tried to describe the PDBs we should create explicitly above.  What further help are you looking for here before this can be implemented?

**Comment by galexrt** [comment]
@anguslees
> What further help are you looking for here before this can be implemented?

Time, more time would be good.

To note here, draining two nodes at the same is a problem in this case as only a `replicated.size: 2` is used which is not really recommended in production (see http://lists.ceph.com/pipermail/ceph-users-ceph.com/2016-September/013237.html and/or ask your local Ceph expert), having `replicated.size: 3` or more is recommended.

**Comment by anguslees** [comment]
> draining two nodes at the same is a problem in this case as only a replicated.size: 2

Yes, but the problem and solution generalises to any level of redundancy.  See the commented-out expressions in the example yaml.

> Time, more time would be good.

So you're ok with the proposal and the next step is implementing the PR?  I can help with that part!
I was concerned the 'help wanted' tag implied you were still uncertain regarding the problem/solution.

**Comment by anguslees** [comment]
Dear stalebot, I appreciate the ongoing value you bring to this issue, please keep up the good work.

**Comment by BlaineEXE** [comment]
> Dear stalebot, I appreciate the ongoing value you bring to this issue, please keep up the good work.

Thanks for your patience and keeping up with this.

**Comment by travisn** [comment]
This was fixed with #3596. The design can be found here: https://github.com/rook/rook/blob/master/design/ceph-managed-disruptionbudgets.md
